src/Common/src/Interop/Windows/mincore/Interop.SHGetKnownFolderPath.cs

```diff
@@ -0,0 +1,358 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Runtime.InteropServices;
+
+internal partial class Interop
+{
+    internal partial class mincore
+    {
+        internal const int COR_E_PLATFORMNOTSUPPORTED = unchecked((int)0x80131539);
+
+        // https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188.aspx
+        [DllImport(Libraries.ShellFolders, CharSet = CharSet.Unicode, SetLastError = false, BestFitMapping = false, ExactSpelling = true, EntryPoint = "SHGetKnownFolderPath")]
+        private static extern int SHGetKnownFolderPath_ShellFolders(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath);
+
+        private static bool s_skipShellFolders;
+
+// Disabling the warning about availability.
+#pragma warning disable BCL0015
+        [DllImport(Libraries.Shell, CharSet = CharSet.Unicode, SetLastError = false, BestFitMapping = false, ExactSpelling = true, EntryPoint = "SHGetKnownFolderPath")]
+        private static extern int SHGetKnownFolderPath_Shell32(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath);
+#pragma warning restore BCL0015
+
+        internal static int SHGetKnownFolderPath(
+            [MarshalAs(UnmanagedType.LPStruct)] Guid rfid,
+            uint dwFlags,
+            IntPtr hToken,
+            out string ppszPath)
+        {
+            // The ShellFolders API set isn't available on Win7, while the direct Shell32 isn't available
+            // on some mincore versions of Windows. We'll attempt the newer API set version and fall back
+            // to the older "Shell32" if needed.
+            if (!s_skipShellFolders)
+            {
+                try
+                {
+                    return SHGetKnownFolderPath_ShellFolders(rfid, dwFlags, hToken, out ppszPath);
+                }
+                catch (TypeLoadException)
```

Comment:
> There is annoying exception thrown that shows up in the debugger. 

Only once and only on Win7.

> It is not compatible with AOT compilation.

Can you explain this more to me? Try-catch around a P/Invoke doesn't work?

> PInvoke to shell32 is not allowed there either (it does not matter that it should not be ever executed).

These folder APIs are actually coming back for UWP 10.1.

Hopefully this is one of the only places we have to do this.

@stephentoub We discussed this briefly yesterday. Do you want to add your two bits?